app/products/calls/state/actions.ts

```diff
@@ -61,6 +70,118 @@ export const setCalls = (serverUrl: string, myUserId: string, calls: Dictionary<
     setCurrentCall(nextCall);
 };
 
+export const processIncomingCalls = async (serverUrl: string, calls: Call[], keepExisting = true) => {
+    if (!getCallsConfig(serverUrl).EnableRinging) {
+        return;
+    }
+
+    const database = DatabaseManager.serverDatabases[serverUrl]?.database;
+    if (!database) {
+        return;
+    }
+
+    // Do we have incoming calls we should notify about?
+    const incomingCalls = getIncomingCalls().incomingCalls;
+    const existingCalls = getCallsState(serverUrl).calls;
+    const myUserId = getCallsState(serverUrl).myUserId;
+    const newIncoming: IncomingCallNotification[] = [];
+
+    for await (const call of calls) {
+        // dismissed already?
+        if (call.dismissed[myUserId] || existingCalls[call.channelId]?.dismissed[myUserId]) {
+            continue;
+        }
+
+        // already in our incomingCalls notifications?
+        if (incomingCalls.findIndex((c) => c.callID === call.id) >= 0) {
+            continue;
+        }
+
+        // Never send a notification for a call you started, or a call you are currently in.
+        if (myUserId === call.ownerId || getCurrentCall()?.id === call.id) {
+            continue;
+        }
+
+        const channel = await getChannelById(database, call.channelId);
+        if (!channel) {
+            logDebug('calls: processIncomingCalls could not find channel by id', call.channelId, 'for serverUrl', serverUrl);
+            continue;
+        }
+
+        if (!isDMorGM(channel)) {
+            continue;
+        }
+
+        const callerModel = await getUserById(database, call.ownerId);
+
+        newIncoming.push({
+            serverUrl,
+            myUserId,
+            callID: call.id,
+            callerID: call.ownerId,
+            callerModel,
+            channelID: call.channelId,
+            startAt: call.startTime,
+            type: channel.type === General.DM_CHANNEL ? ChannelType.DM : ChannelType.GM,
+        });
+    }
+
+    if (newIncoming.length === 0 && keepExisting) {
+        return;
+    }
+
+    if (keepExisting) {
+        newIncoming.push(...incomingCalls);
+    } else {
+        const removedThisServer = incomingCalls.filter((ic) => ic.serverUrl !== serverUrl);
+        newIncoming.push(...removedThisServer);
+    }
+
+    if (newIncoming.length === 0 && incomingCalls.length === 0) {
+        return;
+    }
+
+    newIncoming.sort((a, b) => a.startAt - b.startAt);
+    setIncomingCalls({incomingCalls: newIncoming});
+};
+
+const getChannelIdFromCallId = (serverUrl: string, callId: string) => {
+    const callsState = getCallsState(serverUrl);
+    for (const call of Object.values(callsState.calls)) {
+        if (call.id === callId) {
+            return call.channelId;
+        }
+    }
+    return undefined;
```

Comment:
Is there any practical reason to return `undefined` vs the empty string?

Reply:
Personally, I prefer `undefined` or `null` over using JS's "empty string = falsey" trick/hack/abomination.